includes/classes/shopping_cart.php

```diff
@@ -634,107 +660,104 @@ public function calculate()
         $this->free_shipping_price = 0;
         $this->free_shipping_weight = 0;
         $this->download_count = 0;
-        if (!is_array($this->contents)) return 0;
+        if (!is_array($this->contents)) {
+            return 0;
+        }
 
 // By default, Price Factor is based on Price and is called from function zen_get_attributes_price_factor
 // Setting a define for ATTRIBUTES_PRICE_FACTOR_FROM_SPECIAL to 1 to calculate the Price Factor from Special rather than Price switches this to be based on Special, if it exists
-        if (!defined('ATTRIBUTES_PRICE_FACTOR_FROM_SPECIAL')) define('ATTRIBUTES_PRICE_FACTOR_FROM_SPECIAL', 1);
-        foreach ($this->contents as $products_id => $data) {
+        zen_define_default('ATTRIBUTES_PRICE_FACTOR_FROM_SPECIAL', 1);
+        foreach ($this->contents as $uprid => $data) {
             $total_before_discounts = 0;
-            $freeShippingTotal = $productTotal = $totalOnetimeCharge = $totalOnetimeChargeNoDiscount = 0;
+            $freeShippingTotal = 0;
+            $productTotal = 0;
+            $totalOnetimeCharge = 0;
+            $totalOnetimeChargeNoDiscount = 0;
             $free_shipping_applied = false;
-            $qty = $this->contents[$products_id]['qty'];
+            $qty = $data['qty'];
+            $prid = zen_get_prid($uprid);
 
-            // products price
-            $sql = "SELECT products_id, products_price, products_tax_class_id, products_weight,
-                    products_priced_by_attribute, product_is_always_free_shipping, products_discount_type, products_discount_type_from,
-                    products_virtual, products_model
-                    FROM " . TABLE_PRODUCTS . "
-                    WHERE products_id = " . (int)$products_id;
+            $product = zen_get_product_details($prid);
+            if ($product->EOF) {
+                continue;   //-TODO: Should the product be removed from the cart if this is detected?
+            }
 
-            if ($product = $db->Execute($sql)) {
-                $prid = $product->fields['products_id'];
+            $product = $product->fields;
+            $this->notify('NOTIFY_CART_CALCULATE_PRODUCT_PRICE', $uprid, $product);
+            $prid = (int)$product['products_id'];   //-TODO: Needed?  Just in case the id was changed by the observer?
```

Comment:
Probably good, although it's more likely that other fields would be changed besides the primary key.

Q: While `(int)` is functionally the same as calling `zen_get_prid()`, why not use `zen_get_prid()` for consistency with line 678 above?

Reply:
Update: yes, definitely must do this, for safety, since you've removed the sanitization from the query on line 751 below.